File: doc-tiny/project_contract.py
"""Project contract helpers for doc-tiny.

Mirrors the subset of ``code-tiny/tools/common/project_registry.py`` and
``code-tiny/tools/common/project_scope.py`` that doc-tiny needs. The two
modules cannot import from each other because doc-tiny is laid out as flat
scripts rather than a Python package.

Per Phase 04 / Phase 05 of the unified ingest/query contract plan, every
project-scoped tool in doc-tiny accepts ``project_id`` per call and resolves
its graph + collection through ``resolve_doc_targets``. A missing
``project_id`` falls through to the implicit full-search path (omit it to
query across every project).

The lookup keys are case-insensitive ``casefold()`` values. The naming
contract is:

* ``doc_graph == "{project_id}_doc"`` — separate from the code graph so
  disjoint labels stay queryable.
* ``doc_qdrant_collection == "{project_id}_doc"`` — same convention as
  the code side; ``project_id`` is reused as the suffix.

If a project is registered in ``.cortext-harness/config/*.json``, the
config file's ``doc.env.FALKORDB_GRAPH`` and ``doc.env.QDRANT_COLLECTION``
override the defaults. If no config file exists, the naming rule is
applied.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import Any, Dict, Iterable, List, Mapping, Optional


PROJECT_ID_NORMALIZED_FIELD = "project_id_normalized"


def normalize_project_id(value: Any) -> Optional[str]:
    """Return a non-empty project id, or ``None`` for an unscoped query."""
    if value is None:
        return None
    normalized = str(value).strip()
    return normalized or None


def project_id_lookup_key(value: Any) -> Optional[str]:
    """Return the case-insensitive comparison key for a project id."""
    normalized = normalize_project_id(value)
    return normalized.casefold() if normalized is not None else None


@dataclass(frozen=True)
class ProjectTargets:
    """Doc-side storage targets resolved through the registry."""

    project_id: str
    project_id_normalized: str
    doc_graph: str
    doc_qdrant_collection: str
    source: str = field(default="registry", compare=False)


class ProjectContractError(KeyError):
    """Base class for project-contract errors."""


class ProjectNotRegisteredError(ProjectContractError):
    """Raised when no config describes ``project_id`` and no env seeds it."""

    def __init__(self, project_id: Any, known: Iterable[str]):
        known_list = sorted(known)
        rendered = ", ".join(known_list) if known_list else "<none registered>"
        message = (
            f"project_id '{project_id}' is not registered. "
            f"Known projects: {rendered}. "
            "Add a project section to .cortext-harness/config/*.json or pass "
            "an explicit override."
        )
        super().__init__(message)
        self.project_id = project_id
        self.known = known_list

    def __str__(self) -> str:
        return self.args[0] if self.args else ""


class DuplicateProjectRegistrationError(ProjectContractError):
    """Raised when two descriptors casefold to the same project id."""

    def __init__(self, collisions: Mapping[str, Iterable[str]]):
        self.collisions = {
            key: sorted(str(value) for value in values)
            for key, values in sorted(collisions.items())
        }
        details = "; ".join(
            f"{key}: {', '.join(values)}"
            for key, values in self.collisions.items()
        )
        super().__init__(
            "Duplicate project registrations after casefold normalization: "
            f"{details}."
        )

    def __str__(self) -> str:
        return self.args[0] if self.args else ""


def _default_config_dir() -> Path:
    """Locate ``.cortext-harness/config`` from CWD, walking up."""
    explicit = str(os.environ.get("CORTEX_HARNESS_CONFIG_PATH") or "").strip()
    if explicit:
        path = Path(explicit).expanduser()
        return path if path.is_dir() else path.parent
    cwd = Path.cwd()
    for candidate in (cwd, *cwd.parents):
        path = candidate / ".cortext-harness" / "config"
        if path.is_dir():
            return path
    return cwd / ".cortext-harness" / "config"


def _read_project_entries(config_dir: Path) -> List[Dict[str, Any]]:
    if not config_dir.is_dir():
        return []
    entries: List[Dict[str, Any]] = []
    for file_path in sorted(config_dir.glob("*.json")):
        try:
            with file_path.open(encoding="utf-8") as handle:
                payload = json.load(handle)
        except (OSError, json.JSONDecodeError):
            continue
        if not isinstance(payload, Mapping):
            continue
        project_section = payload.get("project") or {}
        project_id = (
            project_section.get("code")
            or project_section.get("name")
        )
        if not project_id:
            continue
        entries.append(
            {
                "project_id": str(project_id),
                "doc_env": dict(payload.get("doc", {}).get("env") or {}),
            }
        )
    variants: Dict[str, List[str]] = {}
    for entry in entries:
        lookup = project_id_lookup_key(entry.get("project_id"))
        if lookup is not None:
            variants.setdefault(lookup, []).append(str(entry["project_id"]))
    collisions = {key: values for key, values in variants.items() if len(values) > 1}
    if collisions:
        raise DuplicateProjectRegistrationError(collisions)
    return entries


def _resolve_doc_targets(
    project_id: str,
    entries: List[Dict[str, Any]],
) -> ProjectTargets:
    lookup = project_id_lookup_key(project_id)
    if lookup is None:
        raise ProjectNotRegisteredError(project_id, [e["project_id"] for e in entries])
    match: Optional[Dict[str, Any]] = None
    for entry in entries:
        if project_id_lookup_key(entry.get("project_id")) == lookup:
            match = entry
            break
    if match is None:
        raise ProjectNotRegisteredError(
            project_id, [e["project_id"] for e in entries]
        )
    # Canonicalize the raw project_id to the registered form so case
    # variants of the same project collapse to identical ProjectTargets.
    canonical_project_id = str(match["project_id"]) if match is not None else project_id
    doc_env: Mapping[str, Any] = (match or {}).get("doc_env") or {}
    doc_graph = (
        doc_env.get("FALKORDB_GRAPH")
        or doc_env.get("NEO4J_DB")
        or f"{canonical_project_id}_doc"
    )
    doc_qdrant = (
        doc_env.get("QDRANT_COLLECTION")
        or f"{canonical_project_id}_doc"
    )
    return ProjectTargets(
        project_id=canonical_project_id,
        project_id_normalized=lookup,
        doc_graph=doc_graph,
        doc_qdrant_collection=doc_qdrant,
        source="registry",
    )


def resolve_project_targets(
    project_id: Any,
    *,
    config_dir: Optional[Path] = None,
) -> ProjectTargets:
    """Resolve the doc-side storage targets for ``project_id``.

    Raises ``ProjectNotRegisteredError`` if the project is not in any
    config file. Config files must live under
    ``.cortext-harness/config/*.json``.
    """
    normalized = normalize_project_id(project_id)
    if normalized is None:
        raise ProjectNotRegisteredError(project_id, [])
    directory = Path(config_dir) if config_dir is not None else _default_config_dir()
    entries = _read_project_entries(directory)
    return _resolve_doc_targets(normalized, entries)


def list_registered_projects(config_dir: Optional[Path] = None) -> List[str]:
    directory = Path(config_dir) if config_dir is not None else _default_config_dir()
    return [entry["project_id"] for entry in _read_project_entries(directory)]


def with_overrides(targets: ProjectTargets, **overrides: Optional[str]) -> ProjectTargets:
    """Return a new ``ProjectTargets`` with the given fields replaced."""
    valid_fields = {"project_id", "project_id_normalized", "doc_graph", "doc_qdrant_collection"}
    for key in overrides:
        if key not in valid_fields:
            raise ValueError(
                f"unknown ProjectTargets field '{key}'. "
                f"Valid fields: {sorted(valid_fields)}"
            )
    return replace(targets, **overrides)


def qdrant_project_filter(project_id: Any) -> Optional[Dict[str, Any]]:
    """Build the canonical Qdrant payload filter for a doc project scope.

    When ``project_id`` is empty (``None``/blank), the filter is suppressed
    (``None`` returned) so the query crosses project boundaries. This is the
    implicit default for the unified contract.
    """
    normalized = project_id_lookup_key(project_id)
    if normalized is None:
        return None
    return {
        "must": [
            {"key": PROJECT_ID_NORMALIZED_FIELD, "match": {"value": normalized}},
        ],
    }

File: rust/crates/cortex-doc/src/project_contract.rs
//! Port of `doc-tiny/project_contract.py` — the project-registry contract
//! (`{project_id}_doc` naming convention + `.cortext-harness/config/*.json`
//! overrides), including the `CORTEX_HARNESS_CONFIG_PATH` escape hatch.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use serde::Serialize;

pub const PROJECT_ID_NORMALIZED_FIELD: &str = "project_id_normalized";

/// Python `str.casefold()` approximation: full lowercase plus the hard fold
/// of sharp-s (the only casefold mapping that changes string length for the
/// Latin-1/ BMP range project ids use).
pub fn casefold(value: &str) -> String {
    value.to_lowercase().replace('ß', "ss")
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ProjectTargets {
    pub project_id: String,
    pub project_id_normalized: String,
    pub doc_graph: String,
    pub doc_qdrant_collection: String,
}

#[derive(Debug)]
pub enum ContractError {
    NotRegistered {
        project_id: String,
        known: Vec<String>,
    },
    Duplicate {
        message: String,
    },
    Io(String),
}

impl std::fmt::Display for ContractError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ContractError::NotRegistered { project_id, known } => {
                let rendered = if known.is_empty() {
                    "<none registered>".to_string()
                } else {
                    known.join(", ")
                };
                write!(
                    f,
                    "project_id '{project_id}' is not registered. Known projects: {rendered}. \
                     Add a project section to .cortext-harness/config/*.json or pass an \
                     explicit override."
                )
            }
            ContractError::Duplicate { message } => write!(f, "{message}"),
            ContractError::Io(message) => write!(f, "{message}"),
        }
    }
}

impl std::error::Error for ContractError {}

/// `normalize_project_id`.
pub fn normalize_project_id(value: Option<&str>) -> Option<String> {
    let normalized = value?.trim().to_string();
    if normalized.is_empty() {
        None
    } else {
        Some(normalized)
    }
}

/// `project_id_lookup_key`.
pub fn project_id_lookup_key(value: Option<&str>) -> Option<String> {
    normalize_project_id(value).map(|v| casefold(&v))
}

/// `_default_config_dir` — explicit env path, else walk up from CWD looking
/// for `.cortext-harness/config`.
pub fn default_config_dir() -> PathBuf {
    if let Ok(explicit) = std::env::var("CORTEX_HARNESS_CONFIG_PATH") {
        let explicit = explicit.trim().to_string();
        if !explicit.is_empty() {
            let path = PathBuf::from(explicit);
            if path.is_dir() {
                return path;
            }
            return path
                .parent()
                .map(Path::to_path_buf)
                .unwrap_or_else(|| PathBuf::from("."));
        }
    }
    let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let mut candidate = Some(cwd.as_path());
    while let Some(dir) = candidate {
        let path = dir.join(".cortext-harness").join("config");
        if path.is_dir() {
            return path;
        }
        candidate = dir.parent();
    }
    cwd.join(".cortext-harness").join("config")
}

struct RawEntry {
    project_id: String,
    doc_env: BTreeMap<String, String>,
}

/// `_read_project_entries` — parse `*.json` config files sorted by name,
/// collecting `project.code` / `project.name` plus `doc.env` overrides.
fn read_project_entries(config_dir: &Path) -> Result<Vec<RawEntry>, ContractError> {
    let mut entries = Vec::new();
    let Ok(dir_entries) = std::fs::read_dir(config_dir) else {
        return Ok(entries);
    };
    let mut files: Vec<PathBuf> = dir_entries
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().map(|ext| ext == "json").unwrap_or(false))
        .collect();
    files.sort();
    for file_path in files {
        let Ok(content) = std::fs::read_to_string(&file_path) else {
            continue;
        };
        let Ok(payload) = serde_json::from_str::<serde_json::Value>(&content) else {
            continue;
        };
        let Some(project) = payload.get("project") else {
            continue;
        };
        let project_id = project
            .get("code")
            .and_then(serde_json::Value::as_str)
            .or_else(|| project.get("name").and_then(serde_json::Value::as_str));
        let Some(project_id) = project_id else {
            continue;
        };
        let mut doc_env = BTreeMap::new();
        if let Some(env_obj) = payload
            .get("doc")
            .and_then(|doc| doc.get("env"))
            .and_then(serde_json::Value::as_object)
        {
            for (key, value) in env_obj {
                let rendered = match value {
                    serde_json::Value::String(s) => s.clone(),
                    serde_json::Value::Null => continue,
                    other => other.to_string(),
                };
                doc_env.insert(key.clone(), rendered);
            }
        }
        entries.push(RawEntry {
            project_id: project_id.to_string(),
            doc_env,
        });
    }

    // Duplicate registration detection after casefold normalization.
    let mut variants: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for entry in &entries {
        if let Some(lookup) = project_id_lookup_key(Some(&entry.project_id)) {
            variants
                .entry(lookup)
                .or_default()
                .push(entry.project_id.clone());
        }
    }
    let collisions: Vec<(String, Vec<String>)> = variants
        .into_iter()
        .filter(|(_, ids)| ids.len() > 1)
        .collect();
    if !collisions.is_empty() {
        let details = collisions
            .iter()
            .map(|(key, ids)| format!("{key}: {}", ids.join(", ")))
            .collect::<Vec<_>>()
            .join("; ");
        return Err(ContractError::Duplicate {
            message: format!(
                "Duplicate project registrations after casefold normalization: {details}."
            ),
        });
    }
    Ok(entries)
}

fn resolve_targets(
    project_id: &str,
    entries: &[RawEntry],
) -> Result<ProjectTargets, ContractError> {
    let lookup =
        project_id_lookup_key(Some(project_id)).ok_or_else(|| ContractError::NotRegistered {
            project_id: project_id.to_string(),
            known: entries.iter().map(|e| e.project_id.clone()).collect(),
        })?;
    let known = || {
        entries
            .iter()
            .map(|e| e.project_id.clone())
            .collect::<Vec<_>>()
    };
    let matched = entries
        .iter()
        .find(|entry| {
            project_id_lookup_key(Some(&entry.project_id)).as_deref() == Some(lookup.as_str())
        })
        .ok_or_else(|| ContractError::NotRegistered {
            project_id: project_id.to_string(),
            known: known(),
        })?;
    // Canonicalize to the registered spelling.
    let canonical_project_id = matched.project_id.clone();
    let doc_graph = matched
        .doc_env
        .get("FALKORDB_GRAPH")
        .or_else(|| matched.doc_env.get("NEO4J_DB"))
        .cloned()
        .unwrap_or_else(|| format!("{canonical_project_id}_doc"));
    let doc_qdrant_collection = matched
        .doc_env
        .get("QDRANT_COLLECTION")
        .cloned()
        .unwrap_or_else(|| format!("{canonical_project_id}_doc"));
    Ok(ProjectTargets {
        project_id: canonical_project_id,
        project_id_normalized: lookup,
        doc_graph,
        doc_qdrant_collection,
    })
}

/// `resolve_project_targets(project_id)`.
pub fn resolve_project_targets(
    project_id: &str,
    config_dir: Option<&Path>,
) -> Result<ProjectTargets, ContractError> {
    let Some(normalized) = normalize_project_id(Some(project_id)) else {
        return Err(ContractError::NotRegistered {
            project_id: project_id.to_string(),
            known: Vec::new(),
        });
    };
    let directory = config_dir
        .map(Path::to_path_buf)
        .unwrap_or_else(default_config_dir);
    let entries = read_project_entries(&directory)?;
    resolve_targets(&normalized, &entries)
}

/// `resolve_doc_candidates(project_id)` — prefix/LIKE rule over the registry.
pub fn resolve_doc_candidates(
    project_id: &str,
    config_dir: Option<&Path>,
) -> Result<Vec<ProjectTargets>, ContractError> {
    let Some(query_key) = project_id_lookup_key(Some(project_id)) else {
        return Ok(Vec::new());
    };
    let directory = config_dir
        .map(Path::to_path_buf)
        .unwrap_or_else(default_config_dir);
    let entries = read_project_entries(&directory)?;
    let mut matches: Vec<&RawEntry> = entries
        .iter()
        .filter(|entry| {
            project_id_lookup_key(Some(&entry.project_id))
                .map(|key| key.starts_with(&query_key))
                .unwrap_or(false)
        })
        .collect();
    let exact: Vec<&&RawEntry> = matches
        .iter()
        .filter(|entry| {
            project_id_lookup_key(Some(&entry.project_id)).as_deref() == Some(query_key.as_str())
        })
        .collect();
    if !exact.is_empty() {
        matches = exact.into_iter().copied().collect();
    }
    matches
        .iter()
        .map(|entry| resolve_targets(&entry.project_id, &entries))
        .collect()
}

/// `list_registered_projects`.
pub fn list_registered_projects(config_dir: Option<&Path>) -> Result<Vec<String>, ContractError> {
    let directory = config_dir
        .map(Path::to_path_buf)
        .unwrap_or_else(default_config_dir);
    Ok(read_project_entries(&directory)?
        .into_iter()
        .map(|entry| entry.project_id)
        .collect())
}

/// `qdrant_project_filter(project_id, known_ids)` — canonical Qdrant payload
/// filter (JSON shape) for a doc project scope.
pub fn qdrant_project_filter(
    project_id: Option<&str>,
    known_ids: Option<&[String]>,
) -> Option<serde_json::Value> {
    let keys = project_id_scope_keys(project_id, known_ids)?;
    Some(serde_json::json!({
        "must": [
            { "key": PROJECT_ID_NORMALIZED_FIELD, "match": { "any": keys } },
        ],
    }))
}

/// `project_id_scope_keys` — the query's own casefold key plus every known id
/// whose casefold key starts with it.
pub fn project_id_scope_keys(
    project_id: Option<&str>,
    known_ids: Option<&[String]>,
) -> Option<Vec<String>> {
    let query_key = project_id_lookup_key(project_id)?;
    let mut keys = vec![query_key.clone()];
    let known: Vec<String> = match known_ids {
        Some(ids) => ids.to_vec(),
        None => list_registered_projects(None).unwrap_or_default(),
    };
    for id in known {
        if let Some(known_key) = project_id_lookup_key(Some(&id))
            && known_key.starts_with(&query_key)
            && !keys.contains(&known_key)
        {
            keys.push(known_key);
        }
    }
    keys.sort();
    Some(keys)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn write_config(dir: &Path, name: &str, body: &str) {
        std::fs::create_dir_all(dir).unwrap();
        std::fs::write(dir.join(name), body).unwrap();
    }

    #[test]
    fn registry_resolution_and_naming_convention() {
        let tmp = tempfile::tempdir().unwrap();
        write_config(
            tmp.path(),
            "b.json",
            r#"{"project": {"code": "BankA"}, "doc": {"env": {"FALKORDB_GRAPH": "bank_doc"}}}"#,
        );
        write_config(tmp.path(), "a.json", r#"{"project": {"code": "p14doc"}}"#);

        let targets = resolve_project_targets("banka", Some(tmp.path())).unwrap();
        assert_eq!(targets.project_id, "BankA"); // canonical registered spelling
        assert_eq!(targets.project_id_normalized, "banka");
        assert_eq!(targets.doc_graph, "bank_doc"); // doc.env override wins
        assert_eq!(targets.doc_qdrant_collection, "BankA_doc"); // naming convention

        let targets = resolve_project_targets("p14doc", Some(tmp.path())).unwrap();
        assert_eq!(targets.doc_graph, "p14doc_doc");
        assert_eq!(targets.doc_qdrant_collection, "p14doc_doc");

        let list = list_registered_projects(Some(tmp.path())).unwrap();
        assert_eq!(list, vec!["p14doc", "BankA"]); // sorted by file name

        let err = resolve_project_targets("missing", Some(tmp.path())).unwrap_err();
        assert!(err.to_string().contains("is not registered"));
    }

    #[test]
    fn duplicate_registration_detected() {
        let tmp = tempfile::tempdir().unwrap();
        write_config(tmp.path(), "a.json", r#"{"project": {"code": "dup"}}"#);
        write_config(tmp.path(), "b.json", r#"{"project": {"code": "DUP"}}"#);
        let err = resolve_project_targets("dup", Some(tmp.path())).unwrap_err();
        assert!(err.to_string().contains("Duplicate project registrations"));
    }

    #[test]
    fn scope_keys_prefix_expansion() {
        let tmp = tempfile::tempdir().unwrap();
        write_config(
            tmp.path(),
            "a.json",
            r#"{"project": {"code": "bank_android"}}"#,
        );
        write_config(
            tmp.path(),
            "b.json",
            r#"{"project": {"code": "bank_Cplus"}}"#,
        );
        let known = list_registered_projects(Some(tmp.path())).unwrap();
        let keys = project_id_scope_keys(Some("BANK"), Some(&known)).unwrap();
        assert_eq!(keys, vec!["bank", "bank_android", "bank_cplus"]);
        assert!(project_id_scope_keys(None, None).is_none());
    }
}
